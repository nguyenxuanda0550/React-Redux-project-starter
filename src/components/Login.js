import { useState } from 'react';
import { connect } from 'react-redux';
import { Navigate } from 'react-router-dom';
import { authenticateUserWithCredentials } from '../actions/authedUser';

const Login = ({ dispatch, loggedIn }) => {
  const [credentials, setCredentials] = useState({
    username: 'tylermcginnis',
    password: 'abc321',
  });
  const [error, setError] = useState('');

  // Redirect if the user is already logged in
  if (loggedIn) {
    const urlParams = new URLSearchParams(window.location.search);
    const redirectUrl = urlParams.get('redirectTo');
    return <Navigate to={redirectUrl ? redirectUrl : '/'} />;
  }

  // Handle form submission
  const onSubmit = (e) => {
    e.preventDefault();
    const res = dispatch(authenticateUserWithCredentials(credentials));
    if (!res) {
      setError('Invalid username or password');
    }
  };

  // Handle login for predefined users
  const handleLogin = (username, password) => {
    dispatch(authenticateUserWithCredentials({ username, password }));
  };

  return (
    <div className="container d-flex flex-column align-items-center mt-5">
      <h1 className="text-center mb-4">Login</h1>
      <div className="dropdown mb-4">
        <button
          className="btn btn-secondary dropdown-toggle"
          type="button"
          id="dropdownMenuButton"
          data-bs-toggle="dropdown"
          aria-expanded="false"
          data-testid="existing-user-label"
        >
          Existing User
        </button>
        <ul className="dropdown-menu" aria-labelledby="dropdownMenuButton">
          <li>
            <button
              onClick={() => handleLogin('sarahedo', 'password123')}
              className="dropdown-item"
            >
              Sarah Edo
            </button>
          </li>
          <li>
            <button
              onClick={() => handleLogin('tylermcginnis', 'abc321')}
              className="dropdown-item"
            >
              Tyler McGinnis
            </button>
          </li>
          <li>
            <button
              onClick={() => handleLogin('mtsamis', 'xyz123')}
              className="dropdown-item"
            >
              Mike Tsamis
            </button>
          </li>
          <li>
            <button
              onClick={() => handleLogin('zoshikanlu', 'pass246')}
              className="dropdown-item"
            >
              Zenobia Oshikanlu
            </button>
          </li>
        </ul>
      </div>
      <form onSubmit={onSubmit} className="w-100 max-w-sm">
        <div className="mb-3">
          <label htmlFor="username" className="form-label" data-testid="username-label">
            Username
          </label>
          <input
            type="text"
            data-testid="username-input"
            className="form-control"
            value={credentials.username}
            onChange={(e) =>
              setCredentials({
                ...credentials,
                username: e.target.value,
              })
            }
          />
        </div>
        <div className="mb-3">
          <label htmlFor="password" className="form-label" data-testid="password-label">
            Password
          </label>
          <input
            type="password"
            data-testid="password-input"
            className="form-control"
            value={credentials.password}
            onChange={(e) =>
              setCredentials({
                ...credentials,
                password: e.target.value,
              })
            }
          />
        </div>
        {error && (
          <div className="alert alert-danger" role="alert" data-testid="error-message">
            {error}
          </div>
        )}
        <button
          type="submit"
          className="btn btn-primary w-100"
          data-testid="submit-login"
        >
          Login
        </button>
      </form>
    </div>
  );
};

const mapStateToProps = ({ authUser }) => ({
  loggedIn: !!authUser,
});

export default connect(mapStateToProps)(Login);

