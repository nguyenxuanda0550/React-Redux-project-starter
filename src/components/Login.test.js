import React from 'react';
import { configureStore } from '@reduxjs/toolkit';
import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import Login from './Login';
import reducer from '../reducers';

// Configure the Redux store with the provided reducer
const store = configureStore({ reducer });

describe('Login Component', () => {
  test('should render the component', () => {
    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    // Check if the component is rendered and matches the snapshot
    expect(view).toBeDefined();
    expect(view).toMatchSnapshot();
  });

  test('should display all elements', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    // Select elements by their test IDs
    const existingUserLabelElement = screen.getByTestId('existing-user-label');
    const usernameLabelElement = screen.getByTestId('username-label');
    const usernameInputElement = screen.getByTestId('username-input');
    const passwordLabelElement = screen.getByTestId('password-label');
    const passwordInputElement = screen.getByTestId('password-input');
    const submitButtonElement = screen.getByTestId('submit-login');

    // Assert that the text content of the elements is correct
    expect(usernameLabelElement.textContent).toBe('Username');
    expect(passwordLabelElement.textContent).toBe('Password');
    expect(submitButtonElement.textContent).toBe('Login');
    expect(existingUserLabelElement.textContent).toBe('Existing User');

    // Simulate user input
    fireEvent.change(usernameInputElement, { target: { value: 'user' } });
    fireEvent.change(passwordInputElement, { target: { value: 'password' } });

    // Assert that the input values are updated
    expect(usernameInputElement.value).toBe('user');
    expect(passwordInputElement.value).toBe('password');
  });

  test('should display error message when username or password is incorrect', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    // Select elements for interaction
    const usernameInputElement = screen.getByTestId('username-input');
    const passwordInputElement = screen.getByTestId('password-input');
    const submitButtonElement = screen.getByTestId('submit-login');

    // Simulate user input and click on submit button
    fireEvent.change(usernameInputElement, { target: { value: 'user' } });
    fireEvent.change(passwordInputElement, { target: { value: 'password' } });
    fireEvent.click(submitButtonElement);

    // Assert that the error message is displayed
    const errorMessageElement = screen.getByTestId('error-message');
    expect(errorMessageElement.textContent).toBe('Username or password is incorrect');
  });
});
