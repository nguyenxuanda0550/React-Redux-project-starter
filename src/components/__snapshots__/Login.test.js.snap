// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container d-flex flex-column align-items-center mt-5"
      >
        <h1
          class="text-center mb-4"
        >
          Login
        </h1>
        <div
          class="dropdown mb-4"
        >
          <button
            aria-expanded="false"
            class="btn btn-secondary dropdown-toggle"
            data-bs-toggle="dropdown"
            data-testid="existing-user-label"
            id="dropdownMenuButton"
            type="button"
          >
            Existing User
          </button>
          <ul
            aria-labelledby="dropdownMenuButton"
            class="dropdown-menu"
          >
            <li>
              <button
                class="dropdown-item"
              >
                Sarah Edo
              </button>
            </li>
            <li>
              <button
                class="dropdown-item"
              >
                Tyler McGinnis
              </button>
            </li>
            <li>
              <button
                class="dropdown-item"
              >
                Mike Tsamis
              </button>
            </li>
            <li>
              <button
                class="dropdown-item"
              >
                Zenobia Oshikanlu
              </button>
            </li>
          </ul>
        </div>
        <form
          class="w-100 max-w-sm"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              data-testid="username-label"
              for="username"
            >
              Username
            </label>
            <input
              class="form-control"
              data-testid="username-input"
              type="text"
              value="tylermcginnis"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              data-testid="password-label"
              for="password"
            >
              Password
            </label>
            <input
              class="form-control"
              data-testid="password-input"
              type="password"
              value="abc321"
            />
          </div>
          <button
            class="btn btn-primary w-100"
            data-testid="submit-login"
            type="submit"
          >
            Login
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container d-flex flex-column align-items-center mt-5"
    >
      <h1
        class="text-center mb-4"
      >
        Login
      </h1>
      <div
        class="dropdown mb-4"
      >
        <button
          aria-expanded="false"
          class="btn btn-secondary dropdown-toggle"
          data-bs-toggle="dropdown"
          data-testid="existing-user-label"
          id="dropdownMenuButton"
          type="button"
        >
          Existing User
        </button>
        <ul
          aria-labelledby="dropdownMenuButton"
          class="dropdown-menu"
        >
          <li>
            <button
              class="dropdown-item"
            >
              Sarah Edo
            </button>
          </li>
          <li>
            <button
              class="dropdown-item"
            >
              Tyler McGinnis
            </button>
          </li>
          <li>
            <button
              class="dropdown-item"
            >
              Mike Tsamis
            </button>
          </li>
          <li>
            <button
              class="dropdown-item"
            >
              Zenobia Oshikanlu
            </button>
          </li>
        </ul>
      </div>
      <form
        class="w-100 max-w-sm"
      >
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            data-testid="username-label"
            for="username"
          >
            Username
          </label>
          <input
            class="form-control"
            data-testid="username-input"
            type="text"
            value="tylermcginnis"
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            data-testid="password-label"
            for="password"
          >
            Password
          </label>
          <input
            class="form-control"
            data-testid="password-input"
            type="password"
            value="abc321"
          />
        </div>
        <button
          class="btn btn-primary w-100"
          data-testid="submit-login"
          type="submit"
        >
          Login
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
